src/atomate2/common/jobs/anharmonicity.py

```diff
@@ -0,0 +1,611 @@
+    temp: float
+        Temperature in K to find velocity and displacement at
+    rng: np.random.Generator
+        Seeded random number generator
+
+    Returns
+    -------
+    np.ndarray
+        Array with iid normally distributed displacements
+    """
+    n_eigvals = eig_vals.shape[0]
+    spread = np.sqrt(-2.0 * np.log(1.0 - rng.random(size=n_eigvals)))
+
+    # Assign amplitudes (A_s) and phases (phi_s)
+    a_s = spread * (np.sqrt(temp * kb) / eig_vals)
+    phi_s = 2.0 * np.pi * rng.random(size=n_eigvals)
+
+    # Get displacement (not normalized by sqrt(masses) yet)
+    return (a_s * np.cos(phi_s) * eig_vecs).sum(axis=2)
+
+
+@job
+def displace_structure(
+    phonon_supercell: Structure,
+    force_constants: ForceConstants = None,
+    temp: float = 300,
+    one_shot: bool = True,
+    seed: int | None = None,
+    mode_resolved: bool = False,
+    n_samples: int = 1,
+) -> list[Structure]:
+    """Calculate the displaced structure.
+
+    Procedure defined in doi.org/10.1103/PhysRevB.94.075125.
+    Displaced supercell = Original supercell + Displacements
+
+    Parameters
+    ----------
+    phonon_supercell: Structure
+        Supercell to distort
+    force_constants: ForceConstants
+        Force constants calculated from phonopy
+    temp: float
+        Temperature (in K) to displace structure at
+    one_shot: bool
+        If false, uses a normally distributed random number for zeta.
+        The default is true.
+    seed: int | None
+        Seed to use for the random number generator
+        (only used if one_shot_approx == False)
+    mode_resolved: bool
+        If true, calculate the mode-resolved sigma^A. This is false by default.
+    n_samples: int
+        Number of samples to generate (must be >= 1).
+        An error is raised if n_samples == 1 and mode_resolved == True
+
+    Returns
+    -------
+    list[Structure]
+        List of displaced Pymatgen structures
+    """
+    if n_samples != 1 and one_shot:
+        raise ValueError(
+            f"One-shot can't be used unless n_sample == 1 not {n_samples}."
+        )
+    if n_samples < 1:
+        raise ValueError(f"n_sample must be >= 1, not {n_samples}.")
+    if n_samples == 1 and mode_resolved:
+        raise ValueError("n_sample must be > 1 when mode_resolved == True.")
+    rng = np.random.default_rng(seed)
+
+    coords = phonon_supercell.cart_coords
+    disp = np.zeros(coords.shape)
+
+    masses = np.array([site.species.weight for site in phonon_supercell.sites])
+    dynamical_matrix = build_dynmat(force_constants, phonon_supercell)
+
+    eig_val, eig_vec = np.linalg.eigh(dynamical_matrix)
+    eig_val = np.sqrt(eig_val[3:])
+    x_acs = eig_vec[:, 3:].reshape((-1, 3, len(eig_val)))
+
+    # gauge eigenvectors: largest value always positive
+    for ii in range(x_acs.shape[-1]):
+        vec = x_acs[:, :, ii]
+        max_arg = np.argmax(abs(vec))
+        x_acs[:, :, ii] *= np.sign(vec.flat[max_arg])
+
```

Comment:
does it make sense to have this without the workflow as a function somewhere? (not necessary but maybe useful for reuse in the future)

Reply:
Potentially it could be useful, so we should probably move it into a separate function

Reply:
There is already a separate function build_dynmat on line 241. One possibility could be to move lines 210-211 into this function and have it return the eigenvals and eigenvecs. Then this part could just call the updated build_dynmat and gauge the eigenvectors. In the step for mode-resolving sigma^A, build_dynmat is also called but we don't gauge the eigenvectors so I'm not sure if it would be better to have the gauging step separate from the new function

Reply:
I would have the dynamic matrix and the eigenvalue/vector solver because some applications may just need the matrix

Reply:
I separated out the code so that in the jobs/anharmonicity.py file there is one helper function to calculate the dynamical matrix and then a second to diagonalize the dynamical matrix and square root the eigenvalue so that we have the eigenmodes and eigenfrequencies.